File: src/main/java/com/thebluehats/server/game/enchants/PushComesToShove.java
package com.thebluehats.server.game.enchants;

import com.thebluehats.server.game.managers.combat.templates.ArrowDamageTrigger;
import com.thebluehats.server.game.managers.combat.templates.DamageEnchantTrigger;
import com.thebluehats.server.game.managers.combat.templates.EnchantHolder;
import com.thebluehats.server.game.managers.enchants.DamageTriggeredEnchant;
import com.thebluehats.server.game.managers.enchants.EnchantGroup;
import com.thebluehats.server.game.managers.enchants.EnchantProperty;
import com.thebluehats.server.game.managers.enchants.HitCounter;
import com.thebluehats.server.game.managers.enchants.processedevents.DamageEventEnchantData;
import com.thebluehats.server.game.utils.EnchantLoreParser;
import org.bukkit.Material;
import org.bukkit.entity.Arrow;
import org.bukkit.entity.Player;
import org.bukkit.util.Vector;

import javax.inject.Inject;
import java.util.ArrayList;

public class PushComesToShove extends DamageTriggeredEnchant {
    private final EnchantProperty<Float> pctsForce = new EnchantProperty<>(12f, 25f, 35f);
    private final EnchantProperty<Float> damageAmount = new EnchantProperty<>(0f, 1f, 2f);

    private final HitCounter hitCounter;

    @Inject
    public PushComesToShove(HitCounter hitCounter, ArrowDamageTrigger arrowDamageTrigger) {
        super(new DamageEnchantTrigger[] { arrowDamageTrigger });

        this.hitCounter = hitCounter;
    }

    @Override
    public void execute(DamageEventEnchantData data) {
        Player damagee = data.getDamagee();
        int level = data.getLevel();

        Arrow arrow = (Arrow) data.getEvent().getDamager();

        hitCounter.addOne(damagee);

        if (hitCounter.hasHits(damagee, 3)) {
            Vector velocity = arrow.getVelocity().normalize().multiply(pctsForce.getValueAtLevel(level) / 2.35);
            velocity.setY(0);

            damagee.setVelocity(velocity);

            damagee.damage(damageAmount.getValueAtLevel(level));
        }
    }

    @Override
    public String getName() {
        return "Push Comes to Shove";
    }

    @Override
    public String getEnchantReferenceName() {
        return "PCTS";
    }

    @Override
    public ArrayList<String> getDescription(int level) {
        EnchantLoreParser enchantLoreParser = new EnchantLoreParser("Every 3rd shot on a player<br/><aqua>{0}</aqua>");

        enchantLoreParser.addTextIf(level != 1, " and deals <red>{1}</red><br/>extra damage");

        String[][] variables = new String[2][];

        variables[0] = new String[] { "Punch", "Punch V", "Punch VII" };
        variables[1] = new String[] { "", "+0.5❤", "+1❤" };

        enchantLoreParser.setVariables(variables);

        return enchantLoreParser.parseForLevel(level);
    }

    @Override
    public boolean isDisabledOnPassiveWorld() {
        return false;
    }

    @Override
    public EnchantGroup getEnchantGroup() {
        return EnchantGroup.B;
    }

    @Override
    public boolean isRareEnchant() {
        return false;
    }

    @Override
    public Material[] getEnchantItemTypes() {
        return new Material[] { Material.BOW };
    }

    @Override
    public EnchantHolder getEnchantHolder() {
        return EnchantHolder.DAMAGER;
    }
}

File: src/main/java/com/thebluehats/server/game/enchants/PushComesToShove.kt
package com.thebluehats.server.game.enchants

import com.thebluehats.server.game.managers.combat.templates.ArrowDamageTrigger
import com.thebluehats.server.game.managers.combat.templates.DamageEnchantTrigger
import com.thebluehats.server.game.managers.combat.templates.EnchantHolder
import com.thebluehats.server.game.managers.enchants.DamageTriggeredEnchant
import com.thebluehats.server.game.managers.enchants.EnchantGroup
import com.thebluehats.server.game.managers.enchants.EnchantProperty
import com.thebluehats.server.game.managers.enchants.HitCounter
import com.thebluehats.server.game.managers.enchants.processedevents.DamageEventEnchantData
import com.thebluehats.server.game.utils.EnchantLoreParser
import org.bukkit.Material
import org.bukkit.entity.Arrow
import java.util.*
import javax.inject.Inject

class PushComesToShove @Inject constructor(private val hitCounter: HitCounter, arrowDamageTrigger: ArrowDamageTrigger) :
    DamageTriggeredEnchant(arrayOf<DamageEnchantTrigger>(arrowDamageTrigger)) {
    private val pctsForce = EnchantProperty(12f, 25f, 35f)
    private val damageAmount = EnchantProperty(0f, 1f, 2f)
    override fun execute(data: DamageEventEnchantData) {
        val damagee = data.damagee
        val level = data.level
        val arrow = data.event.damager as Arrow
        hitCounter.addOne(damagee)
        if (hitCounter.hasHits(damagee, 3)) {
            val velocity = arrow.velocity.normalize().multiply(pctsForce.getValueAtLevel(level) / 2.35)
            velocity.setY(0)
            damagee.velocity = velocity
            damagee.damage(damageAmount.getValueAtLevel(level).toDouble())
        }
    }

    override fun getName(): String {
        return "Push Comes to Shove"
    }

    override fun getEnchantReferenceName(): String {
        return "PCTS"
    }

    override fun getDescription(level: Int): ArrayList<String> {
        val enchantLoreParser = EnchantLoreParser("Every 3rd shot on a player<br/><aqua>{0}</aqua>")
        enchantLoreParser.addTextIf(level != 1, " and deals <red>{1}</red><br/>extra damage")
        val variables: Array<Array<String>> = arrayOfNulls(2)
        variables[0] = arrayOf("Punch", "Punch V", "Punch VII")
        variables[1] = arrayOf("", "+0.5❤", "+1❤")
        enchantLoreParser.setVariables(variables)
        return enchantLoreParser.parseForLevel(level)
    }

    override fun isDisabledOnPassiveWorld(): Boolean {
        return false
    }

    override fun getEnchantGroup(): EnchantGroup {
        return EnchantGroup.B
    }

    override fun isRareEnchant(): Boolean {
        return false
    }

    override fun getEnchantItemTypes(): Array<Material> {
        return arrayOf(Material.BOW)
    }

    override fun getEnchantHolder(): EnchantHolder {
        return EnchantHolder.DAMAGER
    }
}